README
IO::File::Cached - load on demand caching of files
--------------------------------------------------

The IO::File::Cached module requires the Cache::Cache
module for caching of files. This in turn requires

 Digest::MD5
 IPC::ShareLite
 Storable
 Error
 File::Space


Feedback, bug reports, patches
------------------------------

Please send patches in unified diff format to:

	Daniel Berrange <dan@berrange.com>

-- End


CHANGES
Change Log for Config::Record
=============================

v1.1.1 (on 04-06-2006)
----------------------

This is a minor cleanup release with no backwards compatability
breakage

 * Fixed POD comments
 * Added tests for POD comments & coverage
 * Allow a filename to be passed into save/load methods 
 * Make RPM specfile install into vendor dirs, rather than site

v1.1.0 (on 09-03-2005)
----------------------

This is a new major release since it breaks backwards
compatability with 1.0.x series for compound keys.

 * Allow keys in config files to contain periods.
 * Change syntax for retrieving nested values from being
   period separated to be slash separated. eg, instead of
   $cfg->get("foo.bar"), use $cfg->get("foo/bar");
 * Add support for selecting specific elements in an array
   with $cfg->get("foo/[0]");
 * Add support for creating a view on a subset of the
   configuration record
 * Increased test coverage to hit 83% of statements, 64% of
   branches, 76% of conditionals, and 100% of subroutines.

v1.0.5 (on 10-10-2004)
----------------------

 * Fixed dumb mistake which would cause processing of
   configuration file to stop when encountering a blank
   line.

v1.0.4 (on 26-09-2004)
----------------------

 * Allow comments at the end of lines following a quoted
   string, an open/close bracket, an open/close brace
 * Allow long lines to be split with \ to break lines
 * Allow multi-line strings to be input using <<EOF

v1.0.3 (on 14-05-2004)
----------------------

 * When slurping filehandles only change $/ locally
 * Add 'save' method to write a config file back to disk
 * Add 'set' method for changing a config value
 * Add 'get' method to replace 'param' and print deprecation
   warning for all use of 'param' when 'use warnings' is in
   effect (no warnings 'Config::Record' to disable this).

v1.0.2 (on 05-04-2004)
----------------------

 * Fix stupid mistake of missing 'use IO::File' in Config::Record,
   not caught by test suite.
 * Make RPMs builds portable across Debian & Red Hat by forcing
   man(3) install dir to always be $siteprefix/share/man/man3.

v1.0.1 (on 01-04-2004)
----------------------

 * Fix building of RPMs
 * Remove dependancy on IO::File::Cached & File::Path::Localize in 
   favour of letting application developer use these when required.
 * Split POD out into separate file, since MakeMaker (un)helpfully
   skips any file matching /(config|setup).*\.pm/ :-(

v1.0.0: (12-02-2004)
--------------------

 * Initial release.
  
-- End


Config-Record.spec.PL
# Copyright (C) 2001-2004 Daniel Berrange <dan@berrange.com>
#
# $Id: Config-Record.spec.PL,v 1.8 2006/01/27 16:25:50 dan Exp $

use strict;

die unless (scalar @ARGV == 1);

unless (do 'lib/Config/Record.pm')
  {
    if ($@) { die $@ };
    die "lib/Config/Record.pm: $!";
  }
		
local $/ = undef;
$_ = <DATA>;
s/\@VERSION\@/$Config::Record::VERSION/g;

open SPEC, ">$ARGV[0]" or die "$!";
print SPEC $_;
close SPEC;
__DATA__
# Automatically generated by Config-Record.spec.PL

%define perlvendorarch %(perl -e 'use Config; print $Config{installvendorarch}')
%define perlvendorlib %(perl -e 'use Config; print $Config{installvendorlib}')
%define perlvendorprefix %(perl -e 'use Config; print $Config{vendorprefix}')
%define perlvendorman3 %{perlvendorprefix}/share/man/man3
%define perlversion %(perl -e 'use Config; print $Config{version}')

%define appname Config-Record

Summary: Config::Record - Simple configuration records
Name: perl-%{appname}
Version: @VERSION@
Release: 1
License: GPL
Group: Applications/Internet
Source: %{appname}-%{version}.tar.gz
BuildRoot: /var/tmp/%{appname}-%{version}-root
BuildArchitectures: noarch
Requires: perl >= %{perlversion}

%description
Config::Record provides a module for loading configuration
records. It supports scalar, array and hash parameters nested
to an arbitrary depth.
 
%prep
%setup -q -n %{appname}-%{version}


%build
perl Makefile.PL INSTALLDIRS=vendor PREFIX=$RPM_BUILD_ROOT/usr
make


%install
rm -rf $RPM_BUILD_ROOT
make install INSTALLVENDORMAN3DIR=$RPM_BUILD_ROOT%{perlvendorman3}
find $RPM_BUILD_ROOT -name perllocal.pod -exec rm -f {} \;
find $RPM_BUILD_ROOT -name .packlist -exec rm -f {} \;

%clean
rm -rf $RPM_BUILD_ROOT


%files
%defattr(-,root,root)
%doc AUTHORS
%doc LICENSE
%doc README
%{perlvendorman3}/*
%{perlvendorlib}/Config/Record.pm
%{perlvendorlib}/Config/Record.pod


autobuild.sh
#!/bin/sh

NAME="Config-Record"

set -e

# Make things clean.

make -k realclean ||:
rm -rf MANIFEST blib

# Make makefiles.

perl Makefile.PL PREFIX=$AUTOBUILD_INSTALL_ROOT
make manifest
echo $NAME.spec >> MANIFEST

# Build the RPM.
make
perl -MDevel::Cover -e '' 1>/dev/null 2>&1 && USE_COVER=1 || USE_COVER=0
if [ "$USE_COVER" = "1" ]; then
  cover -delete
  HARNESS_PERL_SWITCHES=-MDevel::Cover make test
  cover
  mkdir blib/coverage
  cp -a cover_db/*.html cover_db/*.css blib/coverage
  mv blib/coverage/coverage.html blib/coverage/index.html
else
  make test
fi

make install

rm -f $NAME-*.tar.gz
make dist

if [ -f /usr/bin/rpmbuild ]; then
  rpmbuild -ta --clean $NAME-*.tar.gz
fi

if [ -f /usr/bin/fakeroot -a -n "$AUTOBUILD_PACKAGE_ROOT" ]; then
  fakeroot debian/rules clean
  fakeroot debian/rules DESTDIR=$AUTOBUILD_PACKAGE_ROOT/debian binary
fi


lib/Config/Record.pm
# -*- perl -*-
#
# Config::Record by Daniel Berrange <dan@berrange.com>
#
# Copyright (C) 2000-2004 Daniel P. Berrange <dan@berrange.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
#
# $Id: Record.pm,v 1.12 2006/01/27 16:25:50 dan Exp $

package Config::Record;

use strict;
use Carp qw(confess cluck);
use IO::File;

use warnings::register;

use vars qw($VERSION);

$VERSION = "1.1.1";

sub new {
    my $proto = shift;
    my $class = ref($proto) || $proto;
    my $self = {};
    my %params = @_;
    
    $self->{record} = exists $params{record} ? $params{record} : {};
    $self->{debug} = $params{debug};
    $self->{filename} = undef;
    
    bless $self, $class;
    
    if (defined $params{file}) {
	$self->load($params{file});
    }
    
    return $self;
}


sub load {
    my $self = shift;
 
    my $file;
    if (@_) {
	$file = shift;
    } elsif ($self->{filename}) {
	$file = $self->{filename};
    } else {
	die "no filename was specified";
    }
    
    my $fh;
    if (ref($file)) {
	if (!$file->isa("IO::Handle")) {
	    confess "file must be an instance of IO::Handle";
	}
	$fh = $file;
    } else {
	$fh = IO::File->new($file)
	    or confess "cannot read from $file: $!";
	$self->{filename} = $file;
    }
    
    local $/ = undef;
    my $data = <$fh>;
    $self->_parse($data,  ref($file) ? "<unknown>" : $file);
    $fh->close 
	or confess "cannot close file: $!";
}


sub _parse {
    my $self = shift;
    my $data = shift;
    my $filename = shift;

    my $value = {};
    my @stack = $value;
    my $here;
    my $continuation;

    my $LABEL = '((?:\w|-|\.)+)';
    my $TRAILING_WHITESPACE = '\s*(?:\#.*)?';
    my $lineno = 0;

    my @lines = split /\n/, $data;

    foreach my $line (@lines) {
	$lineno++;
	warn "$lineno: '$line' '$here' '$continuation'\n" if $self->{debug};
	next if $line =~ m|^\s*#|;
	next if $line =~ m|^\s*$|;

	if ($here) {
	    if ($line =~ /\s*${here}\s*$/) { # EOF
		warn "$lineno: End of here doc\n" if $self->{debug};
		$here = undef;
		$continuation = undef;
	    } else {                # ...
		warn "$lineno: Middle of here doc\n" if $self->{debug};
		${$continuation} .= $line . "\n";
	    }
	} elsif ($continuation) {
	    if ($line =~ /^\s*"(.*?)"\s*(\\)?\s*$/ || # "..."
		$line =~ /^\s*(.*?)\s*(\\)?\s*$/) {   #  ...
		warn "$lineno: Continuation\n" if $self->{debug};
		${$continuation} .= $1;
		$continuation = undef unless $2;
	    } else {
		warn "$lineno: unexpected input '$line'\n";
	    }
	} else {
	    if ($line =~ /^\s*$LABEL\s*=\s*\(${TRAILING_WHITESPACE}$/) { # foo = ( 
		warn "$lineno: Key with array\n" if $self->{debug};
		if (ref($value) eq "ARRAY") {
		    confess "unexpected key,value pair in $filename at line $lineno";
		}
		
		my $key = $1;
		
		my $new = [];
		$value->{$key} = $new;
		$value = $new;
		push @stack, $value;
	    } elsif ($line =~ /^\s*\(${TRAILING_WHITESPACE}$/) { # (
		warn "$lineno: Start of array\n" if $self->{debug};
		if (ref($value) ne "ARRAY") {
		    confess "unexpected array entry in $filename at line $lineno";
		}
		
		my $new = [];
		push @{$value}, $new;
		$value = $new;
		push @stack, $value;
	    } elsif ($line =~ /^\s*\)${TRAILING_WHITESPACE}$/) { # )
		warn "$lineno: End of array\n" if $self->{debug};
		if (ref($value) ne "ARRAY") {
		    confess "mismatched closing round bracket in $filename at line $lineno";
		}
		if ($#stack == 0) {
		    confess "too many closing curley bracket in $filename at line $lineno";
		}
		
		pop @stack;
		$value = $stack[$#stack];
	    } elsif ($line =~ /^\s*$LABEL\s*=\s*{${TRAILING_WHITESPACE}$/) { # foo = {
		warn "$lineno: Key with hash\n" if $self->{debug};
		if (ref($value) eq "ARRAY") {
		    confess "unexpected key,value pair in $filename at line $lineno";
		}
		
		my $key = $1;
		
		my $new = {};
		$value->{$key} = $new;
		$value = $new;
		push @stack, $value;
	    } elsif ($line =~ /^\s*{${TRAILING_WHITESPACE}$/) { # {
		warn "$lineno: Start of hash\n" if $self->{debug};
		if (ref($value) ne "ARRAY") {
		    confess "unexpected array entry in $filename at line $lineno";
		}
		
		my $new = {};
		push @{$value}, $new;
		$value = $new;
		push @stack, $value;
	    } elsif ($line =~ /^\s*}${TRAILING_WHITESPACE}$/) { # }
		warn "$lineno: End of hash\n" if $self->{debug};
		if (ref($value) eq "ARRAY") {
		    confess "mismatched closing curly bracket in $filename at line $lineno";
		}
		if ($#stack == 0) {
		    confess "too many closing curley bracket in $filename at line $lineno";
		}
		
		pop @stack;
		$value = $stack[$#stack];
	    } elsif ($line =~ /^\s*$LABEL\s*=\s*<<(\w+)\s*$/) { # foo = <<EOF
		warn "$lineno: Key with here doc\n" if $self->{debug};
		my $key = $1;
		my $val = "";
		
		$value->{$key} = $val;

		$here = $2;
		$continuation = \$value->{$key};
	    } elsif ($line =~ /^\s*$LABEL\s*=\s*"(.*)"\s*(\\)?${TRAILING_WHITESPACE}$/ || # foo = "..."
		     $line =~ /^\s*$LABEL\s*=\s*(.*?)(\\)?\s*$/) { # foo = ...
		warn "$lineno: Key with string\n" if $self->{debug};
		my $key = $1;
		my $val = $2;
		
		if (ref($value) eq "ARRAY") {
		    confess "expecting value, found key, value pair at line $lineno";
		}
		
		$value->{$key} = $val;
		warn "$lineno: Start continuation\n" if $3 && $self->{debug};
		$continuation = \$value->{$key} if $3;
	    } elsif ($line =~ /^\s*<<(\w+)\s*$/) { # <<EOF
		warn "$lineno: Start of here doc\n" if $self->{debug};
		my $val = "";
		
		if (ref($value) ne "ARRAY") {
		    confess "expecting key,value pair, found value at line $lineno";
		}

		push @{$value}, $val;
		
		$here = $1;
		$continuation = \$value->[$#{$value}];
	    } elsif ($line =~ /^\s*"(.*)"\s*(\\)?${TRAILING_WHITESPACE}$/ || # "..."
		     $line =~ /^\s*(.*?)(\\)?\s*$/) { # ...
		warn "$lineno: Value\n" if $self->{debug};
		my $val = $1;
		
		if (ref($value) ne "ARRAY") {
		    confess "expecting key,value pair, found value at line $lineno";
		}
		
		push @{$value}, $val;
		
		$continuation = \$value->[$#{$value}] if $2;
	    } else {
		warn "Unexpected value '$line'\n";
	    }
	}
    }		 
    if ($#stack != 0) {
	confess "missing closing bracket in $filename at line $lineno";
    }
		 
    $self->{record} = $stack[$#stack];
}
    
sub save {
    my $self = shift;
    
    my $file;
    if (@_) {
	$file = shift;
    } elsif ($self->{filename}) {
	$file = $self->{filename};
    } else {
	die "no filename was specified";
    }
    
    my $fh;
    if (ref($file)) {
	if (!$file->isa("IO::Handle")) {
	    confess "file must be an instance of IO::Handle";
	}
	$fh = $file;
    } else {
	$fh = IO::File->new(">$file")
	    or confess "cannot write to $file: $!";
	$self->{filename} = $file;
    }

    foreach my $key (keys %{$self->{record}}) {
	print $fh "$key = ";
	$self->_format($fh, $self->{record}->{$key}, "");
    }
    
    $fh->close();
}

sub _format {
    my $self = shift;
    my $fh = shift;
    my $value = shift;
    my $indent = shift;

    my $ref = ref($value);
    
    if ($ref) {
	if ($ref eq "HASH") {
	    $self->_format_hash($fh, $value, $indent);
	} elsif ($ref eq "ARRAY") {
	    $self->_format_array($fh, $value, $indent);
	} else {
	    confess "unhandled reference $ref. Configuration files" .
		"can only contain unblessed scalars, array or hash references";
	}
    } else {
	$self->_format_scalar($fh, $value, $indent);
    }
}

sub _format_hash {
    my $self = shift;
    my $fh = shift;
    my $record = shift;
    my $indent = shift;
    
    print $fh "{\n";
    foreach my $key (keys %{$record}) {
	print $fh "$indent  $key = ";
	$self->_format($fh, $record->{$key}, "$indent  ");
    }
    print $fh "$indent}\n";
}

sub _format_array {
    my $self = shift;
    my $fh = shift;
    my $list = shift;
    my $indent = shift;
    
    print $fh "(\n";
    foreach my $element (@{$list}) {
	print $fh "$indent  ";
	$self->_format($fh, $element, "$indent  ");
	
    }
    print $fh "$indent)\n";
}

sub _format_scalar {
    my $self = shift;
    my $fh = shift;
    my $value = shift;
    my $indent = shift;
    
    if ($value =~ /\n/) {
	$value .= "\n" unless $value =~ /\n$/;
	print $fh "<<EOF\n";
	print $fh $value;
	print $fh "EOF\n";
    } elsif ($value =~ /^\s+/ ||
	     $value =~ /\s+$/) {
	# XXX split long lines with \
	# XXX escape embedded "
	print $fh "\"$value\"\n";
    } else {
	# XXX split long lines with \
	print $fh "$value\n";
    }
}


sub view {
    my $self = shift;
    my $key = shift;
    
    my $value = $self->get($key, @_);

    if (!ref($value) ||
	ref($value) ne "HASH") {
	confess "value for $key is not a hash";
    }
    return $self->new(record => $value);
}


sub get {
    my $self = shift;
    my $key = shift;
    
    my @key = split /\//, $key;
    
    my $entry = $self->{record};
    my $context;
    foreach my $fragment (@key) {
	$context = defined $context ? $context . "/" . $fragment : $fragment;
	
	if ($fragment =~ /^\[(\d+)\]$/) {
	    my $index = $1;
	    if (ref($entry) ne "ARRAY") {
		if (@_) {
		    return shift;
		}
		confess "cannot find array value at $context for parameter $key";
	    }
	    if ($#{$entry} < $index) {
		if (@_) {
		    return shift;
		}
		confess "cannot find array value at $context for parameter $key";
	    }
	    $entry = $entry->[$index];
	} elsif ($fragment =~ /((?:\w|-|\.)+)/) {
	    if (ref($entry) ne "HASH") {
		if (@_) {
		    return shift;
		}
		confess "cannot find hash value at $context for parameter $key";
	    }
	    if (!exists $entry->{$fragment}) {
		if (@_) {
		    return shift;
		}
		confess "cannot find hash value at $context for parameter $key";
	    }
	    $entry = $entry->{$fragment};
	} else {
	    confess "fragment '$fragment' should be alphanumeric, or an array index";
	}
    }
    
    return $entry;
}


sub set {
    my $self = shift;
    my $key = shift;
    my $value = shift;
    
    my @key = split /\//, $key;
    
    my $entry = $self->{record};
    my $context;
    while (defined (my $fragment = shift @key)) {
	$context = defined $context ? $context . "/" . $fragment : $fragment;
	
	if ($fragment =~ /^\[(\d+)\]$/) {
	    my $index = $1;
	    if (ref($entry) ne "ARRAY") {
		confess "cannot find array value at $context for parameter $key";
	    }
	    if (@key) {
		if (exists $entry->[$index]) {
		    $entry = $entry->[$index];
		} else {
		    if ($key[0] =~ /^\[(\d+)\]$/) {
			$entry->[$index] = [];
		    } else {
			$entry->[$index] = {};
		    }
		    $entry = $entry->[$index];
		}
	    } else {
		$entry->[$index] = $value;
	    }
	} elsif ($fragment =~ /((?:\w|-|\.)+)/) {
	    if (ref($entry) ne "HASH") {
		confess "cannot find hash value at $context for parameter $key";
	    }
	    if (@key) {
		if (exists $entry->{$fragment}) {
		    $entry = $entry->{$fragment};
		} else {
		    if ($key[0] =~ /^\[(\d+)\]$/) {
			$entry->{$fragment} = [];
		    } else {
			$entry->{$fragment} = {};
		    }
		    $entry = $entry->[$fragment];
		}
	    } else {
		$entry->{$fragment} = $value;
	    }
	} else {
	    confess "fragment '$fragment' should be alphanumeric, or an array index";
	}
    }
}


sub record {
    my $self = shift;
    
    return $self->{record};
}

1 # So that the require or use succeeds.



Makefile.PL
# $Id: Makefile.PL,v 1.3 2006/01/27 16:25:50 dan Exp $
# -*- perl -*-

use ExtUtils::MakeMaker;

WriteMakefile(
	      NAME => "Config::Record",
	      VERSION_FROM => 'lib/Config/Record.pm',
	      AUTHOR => "Daniel P. Berrange (dan[at]berrange[dot]com)",
	      dist =>
	      {
	       COMPRESS => 'gzip --force --best',
	      },

	      clean =>
	      {
	       FILES => '*.bak *~',
	      },
	      
	      depend => {
	        Makefile => '$(VERSION_FROM)',
	        Config-Record.spec => '$(VERSION_FROM)',
		dist => 'Config-Record.spec',
	      },
	      
	      realclean => {
	       FILES => 'Config-Record-*.tar.gz Config-Record.spec',
	      }
	     );

package MY;

sub libscan
  {
    my ($self, $path) = @_;
    ($path =~ /\~$/) ? undef : $path;
  }

sub test {
    my $self = shift;
    my $mm_test = $self->SUPER::test(@_);
    
    return '
TO_TEST_PM = $(TO_INST_PM:lib/%.pm=blib/test/%.pm.tstamp)

test :: test-syntax

test-syntax: pure_all $(TO_TEST_PM)

blib/test/%.pm.tstamp: lib/%.pm
	@echo -n "Checking $<: "
	@perl -I blib/lib -c $<
	@podchecker `echo $< | sed -e \'s/\.pm/.pod/\' `
	@mkdir -p `dirname $@`
	@touch $@

' . $mm_test;
  }

__END__
